torch/csrc/jit/passes/freeze_module.cpp

```diff
@@ -0,0 +1,215 @@
+    auto it = preservedAttrs_.find(&attr_module);
+    return it == preservedAttrs_.end() || !it->second.count(m_name);
+  }
+
+  void recordMutableAttrs(
+      std::shared_ptr<Graph>& graph,
+      const script::Module& module_clone) {
+    std::stack<Block*> blocks({graph->block()});
+
+    while (!blocks.empty()) {
+      Block* block = blocks.top();
+      blocks.pop();
+      for (auto n : block->nodes()) {
+        for (Block* sub_block : n->blocks()) {
+          blocks.push(sub_block);
+        }
+        if (n->kind() == prim::SetAttr) {
+          auto name = n->s(attr::name);
+          auto attr_module = module_clone;
+          auto input = n->inputs()[0]->node();
+          if (!findConstantAttr(input, name, attr_module)) {
+            continue;
+          }
+          insertModifiedAttr(attr_module, name);
+        }
+      }
+    }
+  }
+
+  std::set<std::string> getReferencedAttrs(
+      std::shared_ptr<Graph>& graph,
+      script::Module& module_clone) {
+    std::stack<Block*> blocks({graph->block()});
+    while (!blocks.empty()) {
+      Block* block = blocks.top();
+      blocks.pop();
+      for (auto n : block->nodes()) {
+        for (Block* sub_block : n->blocks()) {
+          blocks.push(sub_block);
+        }
+        if (n->kind() == prim::GetAttr) {
+          auto name = n->s(attr::name);
+          if (module_clone.hasattr(name)) {
+            insertModifiedAttr(module_clone, name);
+          }
+        }
+      }
+    }
+    auto it = preservedAttrs_.find(&module_clone);
+    if (it != preservedAttrs_.end())
+      return it->second;
+    else
+      return std::set<std::string>();
+  }
+
+ public:
+  void propagateAttributes(
+      std::shared_ptr<Graph>& graph,
+      const script::Module& module_clone) {
+    std::stack<Block*> blocks({graph->block()});
+    recordMutableAttrs(graph, module_clone);
+    while (!blocks.empty()) {
+      Block* block = blocks.top();
+      blocks.pop();
+      for (auto it = block->nodes().begin(); it != block->nodes().end();) {
+        Node* n = *it;
+        it++; // advance iterator bc the current node may be destroyed
+
+        for (Block* sub_block : n->blocks()) {
+          blocks.push(sub_block);
+        }
+        if (n->kind() == prim::GetAttr) {
+          WithInsertPoint guard(n);
+          auto name = n->s(attr::name);
+          auto attr_module = module_clone;
+          auto input = n->inputs()[0]->node();
+          if (!findConstantAttr(input, name, attr_module)) {
+            continue;
+          }
+          auto attr = attr_module.attr(name);
+          if (attr.type()->cast<ClassType>()) {
+            continue;
+          }
+
+          if (attr.isTensor()) {
+            auto t = attr.toTensor();
+            t.set_requires_grad(false);
+          }
+          // FIXME: support Dict, TUPLE, genericList and DoubleList constants
+          // to resolve GetAttr.
+          if (attr.isTuple() || attr.isGenericList() || attr.isGenericDict() ||
```

Comment:
What prevents these from being put into constants?

Reply:
This is TODO in the JIT. The types are not yet materializable. Easy future work!